npm/networkPolicyController.go

```diff
@@ -0,0 +1,491 @@
+	}
+
+	if err = c.createCidrsRule("out", netPolObj.ObjectMeta.Name, netPolObj.ObjectMeta.Namespace, egressIPCidrs, ipsMgr); err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: createCidrsRule out due to %v", err)
+	}
+
+	for _, iptEntry := range iptEntries {
+		if err = iptMgr.Add(iptEntry); err != nil {
+			return fmt.Errorf("[syncAddAndUpdateNetPol] Error: failed to apply iptables rule. Rule: %+v with err: %v", iptEntry, err)
+		}
+	}
+
+	cachedNetPolKey := util.GetNSNameWithPrefix(netpolKey)
+	c.npMgr.RawNpMap[cachedNetPolKey] = netPolObj
+
+	metrics.NumPolicies.Inc()
+	// (TODO): may better to use defer?
+	timer.StopAndRecord(metrics.AddPolicyExecTime)
+	return nil
+}
+
+// DeleteNetworkPolicy handles deleting network policy based on cachedNetPolKey.
+func (c *networkPolicyController) deleteNetworkPolicy(netPolObj *networkingv1.NetworkPolicy, isSafeCleanUpAzureNpmChain IsSafeCleanUpAzureNpmChain) error {
+	var err error
+	netpolKey, err := cache.MetaNamespaceKeyFunc(netPolObj)
+	if err != nil {
+		return fmt.Errorf("[deleteNetworkPolicy] Error: while running MetaNamespaceKeyFunc err: %s", err)
+	}
+
+	cachedNetPolKey := util.GetNSNameWithPrefix(netpolKey)
+	cachedNetPolObj, cachedNetPolObjExists := c.npMgr.RawNpMap[cachedNetPolKey]
+	// if there is no applied network policy with the cachedNetPolKey, do not need to clean up process.
+	if !cachedNetPolObjExists {
+		return nil
+	}
+
+	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+	iptMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].iptMgr
+	// translate policy from "cachedNetPolObj"
+	_, _, _, ingressIPCidrs, egressIPCidrs, iptEntries := translatePolicy(cachedNetPolObj)
+
+	// delete iptables entries
+	for _, iptEntry := range iptEntries {
+		if err = iptMgr.Delete(iptEntry); err != nil {
+			return fmt.Errorf("[deleteNetworkPolicy] Error: failed to apply iptables rule. Rule: %+v with err: %v", iptEntry, err)
+		}
+	}
+
+	// delete ipset list related to ingress CIDRs
+	if err = c.removeCidrsRule("in", netPolObj.ObjectMeta.Name, netPolObj.ObjectMeta.Namespace, ingressIPCidrs, ipsMgr); err != nil {
+		return fmt.Errorf("[deleteNetworkPolicy] Error: removeCidrsRule in due to %v", err)
+	}
+
+	// delete ipset list related to egress CIDRs
+	if err = c.removeCidrsRule("out", netPolObj.ObjectMeta.Name, netPolObj.ObjectMeta.Namespace, egressIPCidrs, ipsMgr); err != nil {
+		return fmt.Errorf("[deleteNetworkPolicy] Error: removeCidrsRule out due to %v", err)
+	}
+
+	// (TODO): Can we decouple below logic from cachedNetPolObj since if all deletions for cachedNetPolObj are successful, but UnititNpmChains function is failed,
+	// deleteNetworkPolicy function will be executed again.
+
+	// if there is only one cached network policy in RawNPMap and no immediate network policy to process,
+	// clean up default azure npm chains since the cached network policy from RawNPMap is removed.
+	if isSafeCleanUpAzureNpmChain && len(c.npMgr.RawNpMap) == 1 {
```

Comment:
Instead of this one check, why not move this cleanup operation below delete(c.npMgr.RawNpMap, cachedNetPolKey)

removal of the cache element is critical than the ChainCleanup. Any error in chaincleanup should not block the clean up of cache. Because that will result in more errors.

Reply:
I agree we need to decouple chaincleanup with deletion event, but not sure clean way. It does not cause any more errors. It needs extra computation when retrying happens. If it is ok to leave default Azure NPM chain even though there is no active network policy, we can reverse order. With changing order, we do not have chance to clean up default Azure NPM chain if UninitNpmchains function.

Reply:
One more what overhead is if we leave default Azure chain in data plane side?